add support for heronrc file

prabhuinbarajan opened the pull request:
see readme

kramasamy commented on heron/cli/src/python/.heronrc.sample:
```diff
@@ -0,0 +1,6 @@
+activate:* --config-path cmd-activate-*                   #command specific
```
It will be great if you could rename this file to be heronrc.sample or SAMPLE .heronrc.sample will be hidden in the code and people might not see it.

kramasamy commented:
@prabhuinbarajan - there are some errors. can you please check?

kramasamy commented:
@objmagic - can you please review this?

kramasamy commented:
@taishi8117 @objmagic - can you review this?

objmagic commented:
I will review it as soon as possible on Monday...

taishi8117 commented:
I will do so too

objmagic commented:
Here is my review of this horrible PR. Please bear with me.

#### README.md

First of all, your [README.md](https://github.com/prabhuinbarajan/heron-1/blob/ec5fbfe959aae67dc3f171099aefeb5c887a292c/heron/cli/src/python/README.md) makes little sense to me. What do you mean by "pattern"? I can barely understand what you mean by "precedence"? Do you mean "order"? What is "To run/Test"? What is ``cp.heronrc.sample $HOME/.heronrc``? Do you mean ``cp .heronrc.sample $HOME/.heronrc``?

#### Code

Please fix indentation. We enforce Python indentation level to be 2. There are simply too many parts of your code that need to be fixed, and I don't have time to add specific comment and suggestion to each of them.

#### Test

Please add test under `heron/cli/test/python`. Running test manually is not a viable approach in such a large project.

Twitter's Heron is a high-quality open source project. IMHO, PR like this which shows no professionalism should not be accepted. It needs tons of improvement and further discussion.

objmagic commented:
Since you group config by "precedence", I think it would be better to put each piece of comment on a separate line and stop supporting trailing comment. For example:

````
# command specific
activate:* --config-path cmd-activate-*  

# invalid
:devcluster/ads/PROD --config-path cmd-invalid

# role and command specific
submit:devcluster/ads/PROD --config-path cmd-submit-role  
activate:devcluster/ads/PROD --config-path cmd-activate-role

#global
--config-path hello   
*:* --config-path hello-global

````

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+
+
+if __name__ == "__main__":
+	sys.exit(main())
```
`main` does not return anything. Please add `return 0` at the end of control flow in `main`.

objmagic replied:
why `argparser` module can be invoked as a Python program? If this is for test purpose, please do not run test manually and add test under `heron/cli/tests` to test this module.

objmagic commented on heron/cli/src/python/main.py:
```diff
@@ -25,6 +25,7 @@
 
 import heron.cli.src.python.help as cli_help
 import heron.cli.src.python.args as parse
+import heron.cli.src.python.argparse as hrc
```
Should this be `argparser` instead of `argparse`?

objmagic replied:
Maybe `hrc_parse` is a better module name?

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+
+#test stub
+def main():
+	#parser = HeronRCArgumentParser(description="test parser",fromfile_prefix_chars='@')
```
remove comments

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+						hrc.cmdmap[command][env] =  hrc.cmdmap[command][env] 	+ ' ' + args_list_string	
+					else:
+						hrc.cmdmap[command][env] =  args_list_string
+			#print "cmdmap", hrc.cmdmap	
```
remove commented-out `print`

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+					else:
+						value = hrc.remove_comments(line.rstrip(os.linesep))
+
+					args_list = args.insert_bool_values(value.split())  # make sure that all the single args have a boolean value associated so that we can load the args to a key value structure
```
line too long

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+					m = p.match(line)
+					value = ''
+					command=''
+					env = ''
```
``value, command, env = '', '', ''``

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+import json
+import heron.cli.src.python.activate as activate
+import heron.cli.src.python.args as args
+import traceback
```
please sort import order.

objmagic replied:
remove unnecessary import

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+p = re.compile('(^[^\:]*):([^\s]*) (.*)')
+filters       = ['^@']
+expressions   = [re.compile(x) for x in filters]
+
```
are the above three variables used?

prabhuinbarajan replied:
Yes, these are the HERON_RC defaults

objmagic replied:
`p`, `filters`, `expressions` are not used.

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+					value = ''
+					command=''
+					env = ''
+					if (m != None):
```
`if m is not None:`

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+		if  len(hrc.cmdmap) > 0:
+			return 
+		if os.path.exists(HERON_RC):
+
```
remove unnecessary blank line

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+		super(HeronRCArgumentParser, self).__init__(*args, **kwargs)
+	
+	@classmethod
+	def remove_comments(hrc,string):
```
see my suggestion: https://github.com/twitter/heron/pull/1119#issuecomment-235355454

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python2.7
```
I see no need to run this module as a separate Python program. Please remove this shebang.

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+import argparse
+import os
+import sys
+import shlex
```
remove unnecessary import

objmagic commented on heron/cli/src/python/README.md:
```diff
@@ -0,0 +1,35 @@
+To run/Test : (make sure the __init__.py files are in all places)
+
+cp.heronrc.sample $HOME/.heronrc
+python  -m heron.cli.src.python.argparser activate devcluster/ads/PROD 12313 --config-path "this-is-it"
```
Please use ` to surround the above two commands so that they are formatted properly

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+			print "WARN: %s is not an existing file" % HERON_RC
+
+
+    # for each command / cluster-role-env combination, get the commands from heronrc
```
indent comment properly

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+	def get_args_for_command_role (hrc, command, role):
+		args_for_command_role=''
+
+		if ( command in hrc.cmdmap and role in hrc.cmdmap[command]):
```
unnecessary parenthesis

objmagic commented:
Also, please try `pylint --rcfile=heron/tools/python/checkstyle.ini heron/cli/src/python/argparser.py` to lint your code

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+						hrc.cmdmap[command][env] =  args_list_string
+			#print "cmdmap", hrc.cmdmap	
+		else:
+			print "WARN: %s is not an existing file" % HERON_RC
```
please use Log module under `heron/common` by adding `from heron.common.src.python.color import Log`, then you can use `Log.warn(...)`

objmagic commented on heron/cli/src/python/argparser.py:
```diff
@@ -0,0 +1,143 @@
+	#parser = HeronRCArgumentParser(description="test parser",fromfile_prefix_chars='@')
+	  #parser = argparse.ArgumentParser(
+	parser = HeronRCArgumentParser(
+		prog = 'heron',
```
No space surrounding assignment operator is required. Please write `prog='heron'` like `formatter_class=SubcommandHelpFormatter` below.

prabhuinbarajan commented:
fixed all pylint comments

prabhuinbarajan commented:
@objmagic  - all your code review comments + pylint feedback has been implemented.
@kramasamy  - pls review

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,198 @@
+            cls.cmdmap[app][command][env] = args_list_string
+      Log.debug("RC cmdmap %s", json.dumps(cls.cmdmap))
+    else:
+      Log.warn("WARN: %s is not an existing file", effective_rc)
```
"WARN" is not necessary here. `Log.warn` will add a `Warn` at the beginning.

prabhuinbarajan replied:
complete

objmagic commented on heron/cli/tests/python/argparser_unittest.py:
```diff
@@ -0,0 +1,88 @@
+
+    #suite = unittest.TestSuite()
+    #suite.addTest(HeronRCTest('test_parser_commandline'))
+    #unittest.TextTestRunner(verbosity=2).run(suite)
```
clean up commented out code

prabhuinbarajan replied:
complete

objmagic commented on heron/cli/tests/python/argparser_unittest.py:
```diff
@@ -0,0 +1,88 @@
+
+  def test_parser_commandline(self):
+    sys.argv=[]
+    #print sys.argv
```
clean up this print

prabhuinbarajan replied:
complete

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,198 @@
+          if val is not None and isinstance(val, list):
+            namespace.__dict__[key] = val[0]
+      except Exception:
+        traceback.print_stack()
```
maybe use `Log.debug(traceback.format_exc())`?

prabhuinbarajan replied:
complete and added a warning message as well

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,198 @@
+      except Exception:
+        traceback.print_stack()
+    return namespace, args
+# test stub
```
remove this comment

prabhuinbarajan replied:
complete

objmagic commented:
It looks much better now. I think there are two things left to do:
1. address new comments
2. Please have a look at your [README.md](https://github.com/prabhuinbarajan/heron-1/blob/3f2b67e44350a74e102421ee68b0478f6520d972/heron/cli/src/python/README.md) and fix it with [correct Markdown grammar](https://guides.github.com/features/mastering-markdown/)

kramasamy commented:
@prabhuinbarajan - please address the final set of comments.

objmagic commented:
CI errors because Git merge was not performed correctly. Please fix.

objmagic commented:
you can restart CI on the upper right corner ofTravis CI panel, although I'm not sure if you have the permission

objmagic commented:
1. It seems that `heron` will make using `.heronrc` mandatory? Does it fallback to current parser? Please provide tests for the case that `.heronrc` cannot be found or be parsed correctly. 

2. Address my other comments.

objmagic commented on heron/cli/src/python/.heronrc.sample:
```diff
@@ -0,0 +1,10 @@
+*:*:* --config-property test-hello-global   #global
```
please rename to `heronrc.sample`. we don't want to see this file gets hidden.

prabhuinbarajan replied:
changed to heronrc.sample
the argparser_unittest.py also changed to heronrc.test.

objmagic commented on heron/cli/tests/python/argparser_unittest.py:
```diff
@@ -0,0 +1,85 @@
+      fromfile_prefix_chars='@',
+      add_help = False,
+      rcfile = self.testrcfile,
+      rccommand = "activate",
```
please remove whitespace surrounding assignment operator

prabhuinbarajan replied:
pylint is complaining that it needs one space around assignment. Please advise

pylint --rcfile=tools/python/checkstyle.ini heron/cli/tests/python/argparser_unittest.py

objmagic replied:
you're right. I didn't look carefully.

objmagic commented on heron/cli/tests/python/argparser_unittest.py:
```diff
@@ -0,0 +1,85 @@
+      add_help = False,
+      rcfile = self.testrcfile,
+      rccommand = "activate",
+      rcclusterrole="devcluster/ads/PROD")
```
same as above

prabhuinbarajan replied:
line 63 - 70 addressed

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,197 @@
+        for key in namespace.__dict__:
+          val = namespace.__dict__[key]
+          if val is not None and isinstance(val, list) and len(val) > 0:
+            namespace.__dict__[key] = val[0]
```
can you do `dict_ns = dict(namespace)` first? so you don't have to write `__dict__` everywhere.

prabhuinbarajan replied:
addressed as dict_ns = namespace.__dict__

p.s = dict(namespace) is not working as namespace is not iterable. only namespace.__dict__ is

objmagic replied:
thanks. I misremembered. It should be `vars` (https://github.com/twitter/heron/blob/master/heron/explorer/src/python/main.py#L172)

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,197 @@
+          # structure
+          args_list = config.insert_bool_values(value.split())
+          args_list_string = ' '.join(args_list)
+          if command is None or command == '':
```
equivalent to `if not command:`

prabhuinbarajan replied:
addressed as 
`
if not  command or not app or not env:
            Log.warn("heronrc config entry %s does not have key parameters (command:app:env) ",
                     line)
            continue
`

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,197 @@
+          if command is None or command == '':
+            continue
+          if app is None or app == '':
+            continue
```
equivalent to `if not app:`

prabhuinbarajan replied:
addressed as 
`
if not  command or not app or not env:
            Log.warn("heronrc config entry %s does not have key parameters (command:app:env) ",
                     line)
            continue
`

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,197 @@
+            continue
+          if app is None or app == '':
+            continue
+          if env is None or env == '':
```
equivalent to `if not env:`

prabhuinbarajan replied:
addressed as 
`
if not  command or not app or not env:
            Log.warn("heronrc config entry %s does not have key parameters (command:app:env) ",
                     line)
            continue
`

prabhuinbarajan commented:
@objmagic  - .heronrc is not mandatory. it falls back to the current parser.
added test case :   test_parser_norcfile  in argparser_unittest.py to assert this.
please verify

prabhuinbarajan commented:
all code review comments and style checks  addressed

objmagic commented:
@prabhuinbarajan thanks, I will check this.

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+    if len(cls.cmdmap) > 0:
+      return
+    effective_rc = (rcfile, HERON_RC)[rcfile is None]
+    Log.info('Effective RC file is %s', effective_rc)
```
this should be `Log.debug`

prabhuinbarajan replied:
done

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+# Run the command
+##########################################################################
+
+HERON_RC_FILE = "~/.heronrc"
```
@kramasamy Do you think it's fine to hardcode this file location? I'm OK with this and just want to know your opinion.

kramasamy replied:
@objmagic - it is ok since it is always looking in the user's home directory. 

@prabhuinbarajan - it might be good to change it to 

```os.path.join(os.path.expanduser('~'), '.heronrc')```

prabhuinbarajan replied:
good point, i will combine : HERON_RC_FILE = "~/.heronrc"
HERON_RC = os.path.expanduser(HERON_RC_FILE)

In addition, you may want to note a couple of things:
1. there was supposed to be a system level heron rc (under /etc/)
2. Karthik mentioned that there may be an implicit hierarchy so that at a system level , there could be heron options, and at a user level , all future additions
3. the arg parser currently can also be initialized with an RC file , so the one under home is the default.

objmagic replied:
Do you mean user can specify which rc file to be used on command-line? If so, please add a test for this.

objmagic commented:
@kramasamy My another concern is that user will not be able to know the existence of `.heronrc`, or to say, the possibility of using `.heronrc` to make his life easier. How could user know except reading the source code?

kramasamy commented:
@objmagic - we need to documentation in the web pages. that could be another PR?

prabhuinbarajan commented:
would the current contents of README.md under heron/cli/src/python/ suffice for documentation?

objmagic commented:
do you mean put them as comment/documentation in the source code, or add a way to display them when user enters `help` subcommand?

kramasamy commented:
@prabhuinbarajan and @objmagic - current README is fine. But we need some detailed documentation to be put in the web page. Can you open an issue to track this? We can do the entire documentation in a separate PR.

objmagic commented:
I think this PR is in good shape now. #1204 has just been merged and you need to rebase against master. Specifically, please use log module under `heron/common/src/python/utils`. You can refer to [`heron-explorer`](https://github.com/twitter/heron/blob/master/heron/explorer/src/python/main.py#L31) to see how I config and use this log module.

objmagic commented:
1. Please rebase to resolve conflict with master.
2. Please use the log module --- see the explanation above.

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+      metavar='<command> <options>')
+
+  args, unknown_args = parser.parse_known_args()
+  Log.info("parse results args: %s  unknown: %s ", args, unknown_args)
```
`Log.debug`

prabhuinbarajan replied:
this is the main stub. meant to print it to stdout. info seems okay to me. let me know if you feel strongly

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+          if val is not None and isinstance(val, list) and len(val) > 0:
+            dict_ns[key] = val[0]
+      except Exception:
+        Log.warn("heronrc: unable to clobber arguments (%s,%s ) ", namespace, args)
```
Log.error

prabhuinbarajan replied:
this is not a fatal condition. this means a misconfiguration in heronrc file and hence the warning.

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+# Run the command
+##########################################################################
+# pylint: disable=invalid-name
+Log = log.Log
```
at least in this file, you can do `from import heron.common.src.python.utils.log import Log` since `log` module is not used elsewhere.

prabhuinbarajan replied:
done

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+            cls.cmdmap[app][command][env] = args_list_string
+      Log.debug("RC cmdmap %s", json.dumps(cls.cmdmap))
+    else:
+      Log.warn("%s is not an existing file", effective_rc)
```
If RC file does not exist, we should silently fall back to normal argument parser. That said, should we use `Log.debug` here?

prabhuinbarajan replied:
@objmagic  - HeronRCArgumentParser extends Argparser, and hence the behavior will be defaulting to normal argparser behavior without the file. is that what you mean?

objmagic replied:
right. I think ``Log.error("File %s does not exist", effective_rc)`` is better.

objmagic commented:
and you did not set logging level. See: https://github.com/twitter/heron/blob/master/heron/cli/src/python/main.py#L233

objmagic commented on heron/common/src/python/argparser.py:
```diff
@@ -0,0 +1,196 @@
+# Copyright 2016 Twitter. All rights reserved.
```
I am not sure if you can call it `argparser.py` since it only contains `HeronRCArgumentParser`. Should we rename it to, for example, `rcargparser.py`?

prabhuinbarajan replied:
I am wondering if we should also name the class to be more generic , say HeronArgumentParser instead of making it specific to HeronRC. this also provides the fallback behavior in case the RC file is not found.

objmagic replied:
That's a good point. Please do it.

Kindly,

Runhang

> On Aug 5, 2016, at 08:20, Prabhu Inbarajan <notifications@github.com> wrote:
> 
> In heron/common/src/python/argparser.py:
> 
> > @@ -0,0 +1,196 @@
> > +# Copyright 2016 Twitter. All rights reserved.
> I am wondering if we should also name the class to be more generic , say HeronArgumentParser instead of making it specific to HeronRC. this also provides the fallback behavior in case the RC file is not found.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

prabhuinbarajan replied:
done, changed argparser.py to heronparser.py , 
changed HeronRCArgumentParser to HeronArgumentParser

prabhuinbarajan commented:
@objmagic  - per our conversation this afternoon , added an additional test case for configuring  the equivalent of 
heron submit local --verbose  ~/.heron/examples/heron-examples.jar  com.twitter.heron.examples.ExclamationTopology ExclamationTopology

in heronrc.test as:
````
heron:submit:local --verbose  --topology-file-name ~/.heron/examples/heron-examples.jar  --topology-class-name com.twitter.heron.examples.ExclamationTopology --topology-name ExclamationTopology
`````

also added a baseline test and a comparison test :
test_parser_commandline_norc  (using python's default argparse)
test_parser_commandline_positional_withrc (using argparser )

Please let me know if this meets your expecations. also added some house keeping code for testability in HeronArgumentParser

objmagic commented:
Still this does not work at all.
Here is my `.heronrc`:

````
heron:submit:local --verbose  --topology-file-name ~/.heron/examples/heron-examples.jar  --topology-class-name com.twitter.heron.examples.ExclamationTopology --topology-name ExclamationTopology
````

Neither `heron submit local` nor `heron submit` works for me.

kramasamy commented:
@objmagic - that is not the intended use of heronrc

heron:submit:local --verbose

only work since other options are more positional arguments or rather required arguments.

kramasamy commented on heron/cli/src/python/heronrc.sample:
```diff
@@ -0,0 +1,19 @@
+heron:activate:devcluster/ads/PROD --config-property test-cmd-activate-role
+
+#conversion of positional arguments to optional
+heron:submit:local --verbose  --topology-file-name ~/.heron/examples/heron-examples.jar  --topology-class-name com.twitter.heron.examples.ExclamationTopology --topology-name ExclamationTopology
```
this should not be allowed IMO since positional arguments are always required.

objmagic commented:
LGTM 👍

objmagic commented on heron/common/src/python/heronparser.py:
```diff
@@ -0,0 +1,226 @@
+        for key in dict_ns:
+          val = dict_ns[key]
+          if val is not None and isinstance(val, list) and len(val) > 0:
+            dict_ns[key] = val[0]
```
what's the point of doing this? Can you explain why?

objmagic commented:
related to https://github.com/twitter/heron/issues/1595. I will revert this PR by the end of next Monday if @prabhuinbarajan does not reply or does not want to fix his code.

objmagic commented:
team has reached consensus that we will remove heronrc entirely from codebase
